### ISSUE TITLE
Generate Different Poses

### ISSUE THREAD

**Issue by sunjay** [ProtoArt/spritec#4]
Once we have #3, let's start to automatically set the camera and save the pose to an image. 

A good starting point for this would be to set the camera matrix (the view matrix) to look at the model from 6 different angles (+x, -x, +y, -y, +z, -z) and save each of those as images.

**Comment by sunjay** [comment]
Some more info about this is below.

The code where we compute our MVP matrix is here:

https://github.com/ProtoArt/spritec/blob/a6b5781fc23e7059f46f653d4cfceee63036c680/src/main.rs#L60-L77

Relevant background:

* https://www.evl.uic.edu/ralph/508S98/coordinates.html
* http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/
* https://www.3dgep.com/3d-math-primer-for-game-programmers/
* https://www.3dgep.com/understanding-the-view-matrix/
* http://in2gpu.com/2015/05/17/view-matrix/
* key terms to google: "view matrix", "camera matrix", "view matrix from camera"

The `Mat4<f32>` struct that we're using for transformations has a ton of useful functions. Read through the API docs:

* https://docs.rs/vek/0.9.6/vek/mat/repr_c/column_major/mat4/struct.Mat4.html

We need to figure out a way to do the following:

* [ ] Given a position for the camera and a position to look at in world coordinates, set the `view` variable in the code above to an appropriate matrix that allows the scene to be rendered from that perspective
    * Example: Given the matrix that represents the point `(x=0, y=0, z=20)` and the target point `(x=0, y=0, z=0)`, the camera should render the front of our test model and the entire model should be in view (you may have to play with the `z=20` value a bit)
    * The trouble I'm having with this is that the `Mat4::look_at` function normalizes the vector, so matter what you set the `eye` paramater to, it always stays in the same spot. I have no idea how to deal with this.

Notes:

* Our coordinate system is Y-up, right-handed
* Many of `vek`'s functions assume left-handedness (e.g. `look_at` does this)
* There are private versions of those functions that are right-handed (we can maybe use a fork or just copy the function into our code)
* I am 65% sure our problems are all related to handedness, so make sure you look up each function we're using in that code so you understand what they do
* All the changes you should need to make will be within those lines linked above (feel free to add more lines, of course)

- [ ] We also need to deal with the fact that we are currently scaling the model down. This is *really* bad because scaling down causes the model to lose detail and increases floating point error (as values get closer to zero). Given how low resolution we are, we can't afford to lose even more detail.
- [ ] The first parameter to `perspective_rh_no` is `0.8*PI`. This is our field of view angle. This is a huge angle (out of neccessity because our code doesn't work). Is there a way to make this something more reasonable like `0.3*PI`? I was running into clipping issues and couldn't figure out how to fix them.

The way I suggest going about this is by first removing the call to `scaling_3d` and setting the fov angle to `0.3*PI`, then figuring out how to implement the first checkbox based on your knowledge of matrices and transformations.

**Comment by sunjay** [comment]
The author of vek is fixing the handedness issues, so you should be able to update the dependency shortly. To do that, run `cargo update` and commit the updated Cargo.lock file. 

He's going to make it so there is both a `_lh` and a `_rh` version of look_at. We should use the right handed one because that's our coordinate system. 

Once you use that, the only thing you have to figure out is how to fix the zoom issue. If the camera is farther away, we should be able to see more of the model, but for some reason no matter what I did it always stayed the same. Might have been a handedness issue, so maybe updating vek will fix everything. Crossing my fingers that that's the case.

**Comment by sunjay** [comment]
# Mentoring Instructions

Most of this work was completed as part of #22. We can now generate multiple poses/spritesheets from a configuration file. The part that's left is lots of math!

In essence, all you have to do is:

1. Look through the code for TODO comments that pertain to this issue, read and understand them
2. Resolve each one with the appropriate code

Some more details:

* We now have a `Camera` struct that (currently) just stores the `view` and `projection` matrices. This is not what we want! There is lots more information that we should keep that will allow us to have a camera we can tweak versus a camera we can only set once.

https://github.com/ProtoArt/spritec/blob/70618ca467b74e67b733fe542c81ab2b218749f0/src/shader/camera.rs#L7-L12

* Part of this task is to change `Camera` to store actual information about the camera configuration (the matrices tell us nothing)

https://github.com/ProtoArt/spritec/blob/70618ca467b74e67b733fe542c81ab2b218749f0/src/shader/camera.rs#L58-L72

* The two methods `view` and `projection` will need to be implemented differently to return the right matrices -- these are used in the renderer


* We have a `PresetCamera` that allows the user to either configure from a number of different presets or custom configure their camera to whatever they want
* Part of this task is to implement the `Perspective` presets properly

https://github.com/ProtoArt/spritec/blob/70618ca467b74e67b733fe542c81ab2b218749f0/src/config.rs#L150-L176

* I've cheated and just copied whatever I had in the prototype before and returned those from the `from` method (see also: [The `From` trait](https://doc.rust-lang.org/std/convert/trait.From.html))

https://github.com/ProtoArt/spritec/blob/70618ca467b74e67b733fe542c81ab2b218749f0/src/shader/camera.rs#L14-L56

* You should implement this stuff correctly
* In the future we might want to support more presets for things like orthographic projections

This should hopefully be enough to get you started. You can test your changes by tweaking the configuration. We're using the [TOML](https://github.com/toml-lang/toml) format and the [serde](https://serde.rs/) library to do the parsing/deserialization. It's worthwhile to learn both so you understand the code. You can also ignore all those details and just focus on the math too. :)

Please ask me questions! I am happy to help you test it as well if you are having trouble configuring things. :)

**Comment by sunjay** [comment]
By the way, I wanted to mention that if something in this code seems off, it probably is. I wrote this back when I really didn't know what the future code would look like, so some of my assumptions will definitely turn out to be wrong. :sweat_smile: 

An example of this that I just noticed is that we probably want to pass `width` and `height` fields to the `projection()` method on camera. That way when you do the first bullet under the Mentoring Instructions above, you'll actually be able to calculate the aspect ratio properly. (Our code currently only works for square images.)

```rust
impl Camera {
    // ...

    /// The perspective/orthographic projection of the camera.
    ///
    /// The `width` and `height` parameters are in pixels.
    ///
    /// Camera coordinates -> Homogenous coordinates
    pub fn projection(&self, width: f32, height: f32) -> Mat4<f32> {
        //TODO
    }

    // ...
}
```

I thought I'd leave this comment so you know not to trust my code too much if/when you're getting stuck. If one of my decisions or assumptions makes it hard or impossible to implement something, it's not you, it's me! :sweat_smile: 

Feel free to modify and delete as you see fit until things make sense. :smile:
